4.gene-expression-signatures/scripts/2.run_gsea.py

```diff
@@ -0,0 +1,172 @@
+dependency_df, gene_dict_df= load_model_data(dependency_file, gene_dict_file)
+gene_dict_df = pd.DataFrame(gene_dict_df)
+
+
+# In[3]:
+
+
+#Load weight data for VAEs
+data_directory = pathlib.Path("../0.data-download/data").resolve()
+weight_df = load_train_test_data(
+    data_directory, train_or_test="train"
+)
+
+gene_list_passed_qc = gene_dict_df.loc[
+    gene_dict_df["qc_pass"], "dependency_column"
+].tolist()
+
+weight_data = weight_df.filter(gene_list_passed_qc, axis=1)
+weight_data.head()
+
+
+# In[4]:
+
+
+# Function to extract weights for sklearn models
+def extract_weights(model, model_name):
+    if model_name in ["pca", "ica", "nmf"]:
+        weights_df = pd.DataFrame(model.components_, columns=dependency_df.drop(columns=["ModelID"]).columns.tolist()).transpose()
+        weights_df.columns = [f"{x}" for x in range(0, weights_df.shape[1])]
+    
+    weights_df = weights_df.reset_index().rename(columns={"index": "genes"})
+    return weights_df
+
+# GSEA function (same as before)
+def perform_gsea(weights_df, model_name, num_components, lib="Reactome_2022"):
+    library = blitz.enrichr.get_library(lib)
+    random.seed(0)
+    seed = random.random()
+    gsea_results = []
+    for col in weights_df.columns[1:]:  # Skip 'genes' column
+        gene_signature = weights_df[['genes', col]]
+        if gene_signature.shape[0] > 0:
+            try:
+                gsea_result = blitz.gsea(gene_signature, library, seed=seed)
+                gsea_result = gsea_result.reset_index()
+                for _, pathway_result in gsea_result.iterrows():
+                    result_row = {
+                        "z": int(col),
+                        "full_model_z": num_components,
+                        "model": str(model_name),
+                        "reactome_pathway": str(pathway_result['Term']),
+                        "gsea_es_score": pathway_result['es'],
+                        "nes_score": pathway_result['nes'],
+                        "p_value": pathway_result['pval'],
+                        "shuffled": False
+                    }
+                    gsea_results.append(result_row)
+                    
+            except ZeroDivisionError:
+                print(f"Skipping GSEA for {col} due to zero division error.")
+    
+    gsea_results_df = pd.DataFrame(gsea_results)
+    return gsea_results_df
+
+# Define the location of the saved models and output directory for GSEA results
+model_save_dir = pathlib.Path("saved_models")
+output_dir = pathlib.Path("gsea_results")
+output_dir.mkdir(parents=True, exist_ok=True)
+
+# Latent dimensions and model names to iterate over
+latent_dims = [2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 14, 16, 18, 20, 25, 30, 35, 40, 45, 50, 60, 70, 80, 90, 100, 150, 200]
+model_names = ["pca", "ica", "nmf", "vanillavae", "betavae", "betatcvae"]
+
+final_output_file = output_dir / "combined_z_matrix_gsea_results.parquet"
+try:
+    combined_results_df = pd.read_parquet(final_output_file)
```

Comment:
Double checking: If this file is loaded in the context of this process would it ever be okay to append results (given it'd have data from a prior run)? I wondered if this would lead to potential overlaps or duplicate data somehow.

Reply:
it might, but that shouldn't be a problem! Next PR will be comparing different initializations including this type anyways